### ISSUE TITLE
Proposal: Benchmarking setup

### ISSUE THREAD

**Issue by Perryvw** [TypeScriptToLua/TypeScriptToLua#793]
# Motivation
Benchmarking has been used occasionally as hoc to decide between different solutions for implementing a feature. I would like to provide a way do this in our GitHub environment. This will also allow us to monitor performance degradation or improvement as the project matures.

# Design proposal
I would like to have a benchmarking workflow that can run on GitHub actions and runs for all four of our supported lua versions (not fengari like the tests!). This workflow would have several benchmark scripts that it transpiled to Lua, and then measures the execution time of the Lua over a number of repetitions. I am interested in the **mean execution time** and **standard deviation**.

## Accounting for GitHub actions machine differences
**Assumption: GitHub action host performance stays roughly constant for the duration of one matrix build run.**

We have no control over the way GitHub executes actions and on what machines, so we can not directly compare results between runs. That is a problem. To solve this I propose a 'baseline benchmark', which we can use as benchmark for all other benchmarks in the same build. So results are '1.3x baseline' instead of 1.3s.

# Benchmarking code
I propose adding some benchmark scripts in the `/benchmark` directory in the repository. Their code would be similar to our test builders:
```ts
util.benchmark`
    const add = (x: number, y: number) => x + y;
    return add(1, 2);
`
    . repetitions (100);
```
This code would transpile the TS into lua, and measure the execution time of **only** the Lua execution.

# Further ideas to investigate
We could maybe at some point consider setting some performance gate to display in pull requests. When performance compared to master degrades by a factor X, the run fails.

**Comment by ark120202** [comment]
I don't really get the motivation for this. If that's supposed to help to decide between different implementations, wouldn't you usually want to benchmark pure Lua before writing a transform (or multiple ones, if the feature is not implemented yet)?

I see the need to be able to run benchmarks on all existing lua versions though, but couldn't that just be a bunch of scripts (or a Dockerfile to make it easier to use) outside of the main repo?

**Comment by Perryvw** [comment]
The goal is to have transparent (we can see the benchmarking method), reliable (accurately reflects the performance of the latest tstl result) and reproducible (benchmarking the same code multiple times should yield and results) benchmarks. 

We then have clear results to point at in PRs that we can compare to master to see performance impact of code changes.

Ofcourse you can still do your own benchmarks beforehand to decide on the best transformation for a new feature, but that I think is indeed better to leave out of the repository. (Although maybe we could somehow facilitate that in our CI by also allowing benchmarking raw lua - which should not be merged ofcourse)

**Comment by ark120202** [comment]
Was it resolved by #864, or should we leave this open to track performance benchmarks?

**Comment by lolleko** [comment]
This is partially resolved by #864 If we want to add performance benchmarks we can easily extend our existing (memory) benchmark framework. Although I don't think performance benchmarks will be very useful for us especially if run via GH actions. Either way this should be tracked in another issue.
